Distribution/Server/Features/ReverseDependencies/State.hs

```diff
@@ -13,406 +15,203 @@ import Distribution.PackageDescription
-import Data.SafeCopy (base, deriveSafeCopy)
+import Data.List (union, (\\))
+import Data.Maybe (maybeToList)
+import Data.SafeCopy hiding (Version)
 import Data.Typeable (Typeable)
 import Data.Map (Map)
 import qualified Data.Map as Map
+import Data.Bimap (Bimap, (!), (!>))
+import qualified Data.Bimap as Bimap
+import Control.Applicative
 import Data.Set (Set)
 import qualified Data.Set as Set
-
-import Data.STRef
+import Control.Monad (void)
 import Control.Monad.ST
 import Control.Monad.State (put, get)
-import Control.Monad.Reader (ask, asks)
-
--- The main reverse dependencies map is a drawn-out Map PackageId PackageId,
--- with an extra component to store ranges (all ranges that *could* be revdeps,
--- even if no packages in the index currently satisfy those ranges).
---
--- For selected entries of the map (foo, (2.0, (bar, [1.0]), (bar, (1.0, <3)))):
--- This means that bar-1.0 depends on a version of foo <3, and foo 2.0 meets this criterion.
-type RevDeps = Map PackageName (Map Version (Map PackageName (Set Version)), Map PackageName (Map Version VersionRange))
-
-type CombinedDeps = Map PackageName VersionRange
-
--- TODO: should this be (Maybe (Version, Maybe VersionStatus))?
--- it should be possible, albeit with a bit more work, to determine all revdeps
--- of a package which don't have any versions presently satisfying them.
--- (for an entry (a, b) of RevDeps, take (union a \ b))
-type ReverseDisplay = Map PackageName (Version, Maybe VersionStatus)
+import Control.Monad.Reader (ask)
+import Data.Graph (Graph, Vertex)
+import qualified Data.Graph as Gr
+import qualified Data.Array as Arr ((!), accum, bounds)
+
+type NodeId = Int
+type RevDeps = Graph
 
 data ReverseIndex = ReverseIndex {
-    -- this field is a duplication of PackageIndex, so updates don't have
-    -- to use much information outside of this component, resulting in huge
-    -- happstack-state event files
     duplicatedIndex :: PackageIndex PackageId,
-
-    -- The main reverse dependencies map
     reverseDependencies :: RevDeps,
+    packageNodeIdMap :: Bimap PackageId NodeId
+} deriving (Eq, Show, Typeable)
 
-    -- Generated from packageNameClosure.
-    flattenedReverse :: Map PackageName (Set PackageName),
+instance MemSize ReverseIndex where
+    memSize (ReverseIndex a b c) = memSize3 a b c
 
-    -- Cached counts for reverse dependencies.
-    reverseCount :: Map PackageName ReverseCount
-} deriving (Eq, Show, Typeable)
 
 emptyReverseIndex :: ReverseIndex
-emptyReverseIndex = ReverseIndex (PackageIndex.fromList [])
-                                 Map.empty Map.empty Map.empty
-
-data ReverseCount = ReverseCount {
-    directReverseCount :: Int,
-    flattenedReverseCount :: Int,
-    versionReverseCount :: Map Version Int
-} deriving (Show, Eq, Typeable, Ord)
-emptyReverseCount :: ReverseCount
-emptyReverseCount = ReverseCount 0 0 Map.empty
+emptyReverseIndex = ReverseIndex (PackageIndex.fromList []) (Gr.buildG (0, 20000) []) Bimap.empty
```

Comment:
This 20,000 here is part of the bogus graph doubling thing. It should start completely empty (0,-1) but this involves fixing `addPackage`